packages/tldraw/src/lib/shapes/shared/useAsset.ts

```diff
@@ -1,38 +1,104 @@
-	// more importantly, to not cause zooming in and out to feel janky.
-	const screenScale = useValue('zoom level', () => editor.getZoomLevel() * shapeScale, [
-		editor,
-		shapeScale,
-	])
-
-	useEffect(() => {
-		if (isCulled) return
-
-		let isCancelled = false
-		const timer = editor.timers.setTimeout(async () => {
-			const resolvedUrl = await editor.resolveAssetUrl(assetId, {
-				screenScale,
-			})
-			if (!isCancelled) setUrl(resolvedUrl)
-		}, 500)
-
-		return () => {
-			clearTimeout(timer)
-			isCancelled = true
-		}
-	}, [assetId, asset?.props.src, isCulled, screenScale, editor])
-
-	return { asset, url }
+
+	// The resolved url of the asset that should be used by the shape (duplicated as a ref)
+	const [state, setState] = useState<{ url: string | null; isAnimated: boolean }>({
+		url: null,
+		isAnimated: false,
+	})
+
+	const rPrevUrl = useRef(state.url)
+	// The previous screen scale at which we requested a new url
+	const rPrevScreenScale = useRef(0)
+	// A timeout we can clear if we need to request a new url
+	const rTimeout = useRef<any>(-1)
+
+	useReactor(
+		'update url',
+		() => {
+			const shape = editor.getShape<TLImageShape | TLVideoShape>(shapeId)
+			if (!shape) return
+
+			const asset = assetId ? editor.getAsset(assetId) : null
+			if (!asset) return
+
+			if (editor.getCulledShapes().has(shapeId)) return
+
+			const screenScale =
+				editor.getZoomLevel() * ('w' in asset.props ? shape.props.w / asset.props.w : 1)
+
+			// If the prev scale is zero (which is impossible), set it to the current
+			// screen scale and immediately request a new resolved url.
+			if (!rPrevScreenScale.current) {
+				rPrevScreenScale.current = screenScale
+				editor
+					.resolveAssetUrl(assetId, {
+						screenScale,
+					})
+					.then((resolvedUrl) => {
+						rPrevUrl.current = resolvedUrl
+						setState({
+							url: resolvedUrl,
+							isAnimated:
+								('mimeType' in asset.props &&
+									MediaHelpers.isAnimatedImageType(asset?.props.mimeType)) ||
+								('isAnimated' in asset.props && asset.props.isAnimated),
+						})
+					})
+				return
+			}
+			// If the camera is still moving, don't get a new url
+			if (editor.getCameraState() !== 'idle') return
+
+			// If the editor is not in the idle state (ie if it is resizing) don't get a new url
+			if (!editor.isIn('select.idle')) return
+
+			// If the current screen scale is the same as it was last time we got a new url,
+			// then we don't need to ask for a new one.
+			if (rPrevScreenScale.current === screenScale) return
+			rPrevScreenScale.current = screenScale
+
+			// Clear any previous timers and set a new one. If we make it here again before the
+			// request is completed, we will clear this timer and set a new one. There's a chance
+			// that the timeout is cancelled after the timeout runs but before the promise resolves,
+			// but we'll just ignore that case (maybe a cancellable promise would be the solution).
```

Comment:
hm, i don't know if we can ignore this. it'll get into potentially some weird concurrency bugs. cancellable promise sounds good or via some other mechanism.